mitmproxy/proxy/layers/quic.py

```diff
@@ -0,0 +1,1052 @@
+        if peer_quic is None:
+            (self.buffer_from_client if from_client else self.buffer_from_server).append(event)
+            return
+        peer_connection = self.context.server if from_client else self.context.client
+
+        if isinstance(event, quic_events.DatagramFrameReceived):
+            # forward datagrams (that are not stream-bound)
+            if self.flow is not None:
+                udp_message = udp.UDPMessage(from_client, event.data)
+                self.flow.messages.append(udp_message)
+                yield UdpMessageHook(self.flow)
+                data = udp_message.content
+            else:
+                data = event.data
+            peer_quic.send_datagram_frame(data)
+
+        elif isinstance(event, quic_events.StreamDataReceived):
+            # ignore data received from already ended streams
+            stream = yield from self.get_or_create_stream(event.stream_id)
+            if stream.has_ended(from_client):
+                yield commands.Log(f"Received {len(event.data)} byte(s) on already closed stream #{event.stream_id}.", level="debug")
+                return
+
+            # forward the message allowing addons to change it
+            if stream.flow is not None:
+                tcp_message = tcp.TCPMessage(from_client, event.data)
+                stream.flow.messages.append(tcp_message)
+                yield TcpMessageHook(stream.flow)
+                data = tcp_message.content
+            else:
+                data = event.data
+            peer_quic.send_stream_data(
+                stream.stream_id,
+                data,
+                event.end_stream,
+            )
+
+            # mark the stream as ended if needed
+            if event.end_stream:
+                yield from stream.mark_ended(from_client)
+
+        elif isinstance(event, quic_events.StreamReset):
+            # ignore resets from already ended streams
+            stream = yield from self.get_or_create_stream(event.stream_id)
+            if stream.has_ended(from_client):
+                yield commands.Log(f"Received reset for already closed stream #{event.stream_id}.", level="debug")
+                return
+
+            # forward resets to peer streams and report them to addons
+            peer_quic.reset_stream(
+                stream.stream_id,
+                event.error_code,
+            )
+
+            # mark the stream as failed
+            yield from stream.mark_ended(from_client, err=error_code_to_str(event.error_code))
+
+        else:
+            # ignore other QUIC events
+            yield commands.Log(f"Ignored QUIC event {event!r}.", level="debug")
+            return
+
+        # transmit data to the peer
+        yield QuicTransmit(peer_connection)
+
+    @expect(events.Start)
+    def state_start(self, _) -> layer.CommandGenerator[None]:
+        # mark the main flow as started
+        if self.flow is not None:
+            yield UdpStartHook(self.flow)
```

Comment:
Do I get this right that there's a fake top level "UDP flow" that is only use for signalling?

Reply:
And datagrams :)